cmd/data.go

```diff
@@ -112,77 +114,110 @@ func (cmd *DataCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface
 			return subcommands.ExitUsageError
 		}
 		if targetProfile.TargetDb != "" && conv.TargetDb != targetProfile.TargetDb {
-			err = fmt.Errorf("running data migration for Spanner dialect: %v, whereas schema mapping was done for dialect: %v", targetProfile.TargetDb, conv.TargetDb)
+			fmt.Printf("running data migration for Spanner dialect: %v, whereas schema mapping was done for dialect: %v\n", targetProfile.TargetDb, conv.TargetDb)
 			return subcommands.ExitUsageError
 		}
 	}
 
-	adminClient, err := utils.NewDatabaseAdminClient(ctx)
-	if err != nil {
-		err = fmt.Errorf("can't create admin client: %w", utils.AnalyzeError(err, dbURI))
-		return subcommands.ExitFailure
-	}
-	defer adminClient.Close()
-
-	// Populate migration request id and migration type in conv object
-	conv.Audit.MigrationRequestId = "HB-" + uuid.New().String()
-	conv.Audit.MigrationType = migration.MigrationData_DATA_ONLY.Enum()
+	if !cmd.dryRun {
+		now := time.Now()
+		project, instance, dbName, err := targetProfile.GetResourceIds(ctx, now, sourceProfile.Driver, ioHelper.Out)
+		if err != nil {
+			return subcommands.ExitUsageError
+		}
+		fmt.Println("Using Google Cloud project:", project)
+		fmt.Println("Using Cloud Spanner instance:", instance)
+		utils.PrintPermissionsWarning(sourceProfile.Driver, ioHelper.Out)
 
-	if !sourceProfile.UseTargetSchema() {
-		err = conversion.CreateOrUpdateDatabase(ctx, adminClient, dbURI, sourceProfile.Driver, targetProfile.TargetDb, conv, ioHelper.Out)
+		dbURI := fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName)
+		client, err := utils.GetClient(ctx, dbURI)
 		if err != nil {
-			err = fmt.Errorf("can't create/update database: %v", err)
+			fmt.Printf("can't create client for db %s: %v\n", dbURI, err)
 			return subcommands.ExitFailure
 		}
-	}
+		defer client.Close()
 
-	dataCoversionStartTime := time.Now()
-	bw, err := conversion.DataConv(ctx, sourceProfile, targetProfile, &ioHelper, client, conv, true, cmd.writeLimit)
-	if err != nil {
-		err = fmt.Errorf("can't finish data migration: %v", err)
-		return subcommands.ExitFailure
-	}
-	dataCoversionEndTime := time.Now()
-	dataCoversionDuration := dataCoversionEndTime.Sub(dataCoversionStartTime)
-	conv.Audit.DataConversionDuration = dataCoversionDuration
+		adminClient, err := utils.NewDatabaseAdminClient(ctx)
+		if err != nil {
+			fmt.Printf("can't create admin client: %v\n", utils.AnalyzeError(err, dbURI))
+			return subcommands.ExitFailure
+		}
+		defer adminClient.Close()
+		if !sourceProfile.UseTargetSchema() {
+			dbExists, err := conversion.CheckExistingDb(ctx, adminClient, dbURI)
+			if err != nil {
+				fmt.Printf("can't verify target database: %v\n", err)
+				return subcommands.ExitFailure
+			}
+			if !dbExists {
+				fmt.Println("target database doesn't exist")
+				return subcommands.ExitFailure
+			}
+			err = conversion.ValidateTables(ctx, client, conv.TargetDb)
+			if err != nil {
+				fmt.Printf("can't validate the tables: %v\n", err)
```

Comment:
nit: If we return an error saying "table xyz has non-zero rows", the whole message would read: 
`can't validate the tables: table xyz has non-zero rows`.
Consider rephrasing this error to: `error validating tables`